content/post/2020-03-05-user-preferences/index.Rmd

```diff
@@ -0,0 +1,254 @@
+As [explained by Gábor Csárdi in an R-pkg-devel email](https://www.mail-archive.com/r-package-devel@r-project.org/msg02460.html), _"Applications can actually store user level configuration information, cached data, logs, etc. in the user's home directory, and there standard way to do this [depending on the operating system]."_
+R packages that are on CRAN cannot write to the home directory without warning, but they can and should use standard locations.
+To find where those are, package developers can use the [`rappdirs` package](https://github.com/r-lib/rappdirs).
+
+```{r rappdirs}
+# Using an R6 object
+rhub_app <- rappdirs::app_dir("rhub", "r-hub")
+rhub_app$cache()
+# or functions
+rappdirs::user_cache_dir("rhub")
+```
+
+On top of these non-R specific standard locations, we'll also mention the standard homes of R _options_ and _environment variables_, .Rprofile and .Renviron.
+
+## User preferences
+
+As written in [Android developer guidance](https://developer.android.com/training/id-auth/identify) and probably every customer service guide ever, _"Everyone likes it when you remember their name"_. 
+Everyone probably likes it too when the barista at their favourite coffee shop remembers their usual order. 
+As an R package developer, what can you do for your R package to correctly assess user preferences and settings?
+
+### Using options
+
+In R, `options` _allow the user to set and examine a variety of global options which affect the way in which R computes and displays its results_. For instance, for the usethis package, the [`usethis.quiet` option can control whether usethis is chatty](https://github.com/r-lib/usethis/blob/c0acd1b5e43f03773a5934c1d937bf70b77a2557/principles.md#communicating-with-the-user)[^1]. Users either:
+
+* write `options(usethis.quiet = TRUE)` at the beginning of a script or directly in the console;
+
+* or write that same line in their [.Rprofile that's loaded before every session](https://rstats.wtf/r-startup.html), which is more persistent. 
+
+Users can use a project-level or more global user-level .Rprofile. 
+The use of a project-level .Rprofile overrides the user-level .Rprofile unless the project-level .Rprofile contains the following lines [as mentioned in the `blogdown` book](https://bookdown.org/yihui/blogdown/global-options.html):
+
+```r
+# in .Rprofile of the project
+if (file.exists('~/.Rprofile')) {
+  base::sys.source('~/.Rprofile', envir = environment())
+}
+# then set project options
+```
+
+For more startup tweaks, the user could adopt [the `startup` package](https://cran.r-project.org/web/packages/startup/index.html).
+
+As a package developer in your code you can retrieve options by using `getOption()` whose second argument is a fallback for when the option hasn't been set by the user.
+Note that an option can be any R object.
+
+```{r options}
+options(blabla.foo = TRUE)
+if (getOption("blabla.foo", FALSE)) {
```

Comment:
Because the option can be any object, it is good practice to write `isTRUE(getOption(...))`. If you want to mention this...